src/ol_scroll_window.c

```diff
@@ -295,13 +295,7 @@ ol_scroll_window_set_whole_lyrics (OlScrollWindow *scroll,
     g_ptr_array_ref (whole_lyrics);
     priv->saved_lrc_y = -1;
   }
-  else
-  {
-    /* We only queue draw when the lyrics are no available.
```

Comment:
To be honest I don't remember why I added this. I read the code again and think the reason may be track changing or display mode changing:
https://github.com/osdlyrics/osdlyrics/blob/9894cbc5f000c8fa468996ecc298ec264e424171/src/ol_main.c#L322-L332
https://github.com/osdlyrics/osdlyrics/blob/9894cbc5f000c8fa468996ecc298ec264e424171/src/ol_main.c#L150-L173

In these two cases, the caller doesn't set played time after setting the lyrics. So there may be a chance that the lyrics are rendered before the correct time is set, causing noticeable glitches when the correct time is set.

For mode changing it's easy, just cache the played time and set it right away. For track changes, it's not guaranteed it starts from 0, but most players might start from 0 so it's likely ok to set played time to 0 there.

Reply:
I think I had reached more or less the same conclusions. Implementing your suggestion above should solve this problem: refreshing is explicitly done from ol_main, and there's no guessing work needed in `set_lrc()`.